File: dailyQuestion/2020-06/06-11/golang/solution_stack.go
package main

import "fmt"

func main() {

	T := []int{73, 74, 75, 71, 69, 72, 76, 73}
	result := dailyTemperatures(T)
	fmt.Println(result)

}

func dailyTemperatures(T []int) []int {
	n := len(T)
	result := make([]int, n)
	stack := []int{}
	for i := 0; i < n; i++ {
		for len(stack) != 0 && T[stack[len(stack)-1]] < T[i] {
			result[stack[len(stack)-1]] = i - stack[len(stack)-1]
			stack = stack[:len(stack)-1]
		}
		stack = append(stack, i)
	}
	return result
}

File: dailyQuestion/2020-06/06-11/python/solution_stack.py
# 单调栈
def dailyTemperatures(T: [int]) -> [int]:
    n = len(T)
    result = [0] * n
    stack = []
    for i in range(n):
        while stack and T[stack[-1]] < T[i]:
            tmp = stack.pop()
            result[tmp] = i - tmp
        stack.append(i)
    return result

if __name__ == "__main__" :
    T = [73,74,75,71,69,72,76,73]
    result = dailyTemperatures(T)
    print(result)